digdag-plugin-utils/src/main/java/io/digdag/util/Md5CountInputStream.java

```diff
@@ -0,0 +1,106 @@
+package io.digdag.util;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class Md5CountInputStream
+        extends FilterInputStream
+{
+    private static final MessageDigest MD5;
+
+    static {
+        try {
+            MD5 = MessageDigest.getInstance("MD5");
+        }
+        catch (NoSuchAlgorithmException ex) {
+            throw new RuntimeException(ex);
+        }
+    }
+
+    public static byte[] digestMd5(byte[] data)
+    {
+        try {
+            MessageDigest md5 = (MessageDigest) MD5.clone();
+            return md5.digest(data);
+        }
+        catch (CloneNotSupportedException ex) {
+            throw new RuntimeException("Failed to initialize MD5 digest", ex);
+        }
+    }
+
+    private final MessageDigest md5;
+    private long count;
+
+    public Md5CountInputStream(InputStream in)
+    {
+        super(in);
+        try {
+            this.md5 = (MessageDigest) MD5.clone();
+        }
+        catch (CloneNotSupportedException ex) {
+            throw new RuntimeException("Failed to initialize MD5 digest", ex);
+        }
+        this.count = 0L;
+    }
+
+    public long getCount()
+    {
+        return count;
+    }
+
+    @Override
+    public int read() throws IOException
+    {
+        int c = super.read();
+        if (c >= 0) {
+            count += 1;
+            md5.update((byte) c);
+        }
+        return c;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException
+    {
+        int n = super.read(b, off, len);
+        if (n > 0) {
+            count += n;
+            md5.update(b, off, n);
+        }
+        return n;
+    }
+
+    public int read(byte[] b) throws IOException
+    {
+        return read(b, 0, b.length);
+    }
+
+    @Override
+    public long skip(long n) throws IOException
+    {
+        return read(new byte[(int) Math.min(n, 64*1024L)]);
+    }
+
+    @Override
+    public int available() throws IOException
+    {
+        return in.available();
+    }
+
+    @Override
+    public boolean markSupported()
+    {
+        return false;
+    }
+
+    @Override
+    public void mark(int readlimit)
+    { }
```

Comment:
According to [javadoc](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html), mark should do nothing if markSupported returns false. But on the other hand, reset should throw IOException. This is the behavior of FileInputStream backed by a pipe:

```ruby
irb(main):009:0> fin = java.io.FileInputStream.new(java.io.FileDescriptor.in)
=> #<Java::JavaIo::FileInputStream:0x1cab0bfb>
irb(main):010:0> fin.mark(1024)
=> nil
irb(main):011:0> fin.reset
Java::JavaIo::IOException: mark/reset not supported
	from java.io.InputStream.reset(java/io/InputStream.java:348)
	from java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)
	from RUBY.<eval>((irb):11)
	from org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:976)
	from org.jruby.RubyKernel.loop(org/jruby/RubyKernel.java:1287)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1096)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1096)
	from Users.frsyuki.$_dot_rbenv.versions.jruby_minus_9_dot_0_dot_5_dot_0.bin.irb.<top>(/Users/frsyuki/.rbenv/versions/jruby-9.0.5.0/bin/irb:13)
	from java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:627)
irb(main):012:0> fin.markSupported
=> false
```